src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -100,46 +183,82 @@ func (s *commitLogSource) ReadData(
 		return result.NewDataBootstrapResult(), nil
 	}
 
-	readCommitLogPredicate := newReadCommitLogPredicate(
-		ns, shardsTimeRanges, s.opts, s.inspection)
+	var (
+		fsOpts         = s.opts.CommitLogOptions().FilesystemOptions()
+		filePathPrefix = fsOpts.FilePathPrefix()
+	)
 
-	// TODO(rartoul): When we implement caching data across namespaces, this will need
-	// to be commitlog.ReadAllSeriesPredicate() if CacheSeriesMetadata() is enabled
-	// because we'll need to read data for all namespaces, not just the one we're currently
-	// bootstrapping.
-	readSeriesPredicate := newReadSeriesPredicate(ns)
-	iterOpts := commitlog.IteratorOpts{
-		CommitLogOptions:      s.opts.CommitLogOptions(),
-		FileFilterPredicate:   readCommitLogPredicate,
-		SeriesFilterPredicate: readSeriesPredicate,
+	// Determine which snapshot files are available.
+	snapshotFilesByShard, err := s.snapshotFilesByShard(
+		ns.ID(), filePathPrefix, shardsTimeRanges)
+	if err != nil {
+		return nil, err
 	}
+
+	var (
+		bOpts     = s.opts.ResultOptions()
+		blOpts    = bOpts.DatabaseBlockOptions()
+		blockSize = ns.Options().RetentionOptions().BlockSize()
+	)
+
+	// Determine the minimum number of commit logs files that we
+	// must read based on the available snapshot files.
+	readCommitLogPred, mostRecentCompleteSnapshotByBlockShard, err := s.newReadCommitLogPredBasedOnAvailableSnapshotFiles(
+		ns, shardsTimeRanges, snapshotFilesByShard)
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup the commit log iterator.
+	var (
+		nsID              = ns.ID()
+		seriesSkipped     int
+		datapointsSkipped int
+		datapointsRead    int
+
+		// TODO(rartoul): When we implement caching data across namespaces, this will need
+		// to be commitlog.ReadAllSeriesPredicate() if CacheSeriesMetadata() is enabled
+		// because we'll need to read data for all namespaces, not just the one we're currently
+		// bootstrapping.
+		readSeriesPredicate = func(id ident.ID, namespace ident.ID) bool {
+			shouldReadSeries := nsID.Equal(namespace)
```

Comment:
Hm bugger, yeah this is a pain to have to check for each datapoint.  There's no easy way to make this an integer check or something with some negotiation function?  I guess we'd have to change the actual file format to be an int + namespace string. Then if you receive an int that you haven't seen before you insert it as a new namespace in your callback map or something.

Reply:
Are you referring to do the TODO (which will require some difficult juggling to match data by namespace without doing an expensive bytes comparison), or the existing readSeriesPredicate? The existing readSeriesPredicate is efficient because it gets pushed down to the commit log reader and cached in the metadata map so we only have to do it once per series instead of once per datapoint

Reply:
The TODO for avoiding expensive bytes comparison.  Yeah, makes sense.